native/android/src/main/java/com/nozbe/watermelondb/DatabaseBridge.kt

```diff
@@ -6,93 +6,181 @@ import com.facebook.react.bridge.ReactContextBaseJavaModule
 import com.facebook.react.bridge.ReactMethod
 import com.facebook.react.bridge.Promise
 import com.facebook.react.bridge.ReadableArray
+import com.facebook.react.bridge.Arguments
 import com.nozbe.watermelondb.DatabaseDriver.Operation
 
 class DatabaseBridge(private val reactContext: ReactApplicationContext) :
         ReactContextBaseJavaModule(reactContext) {
 
-    private val connections: MutableMap<ConnectionTag, DatabaseDriver> = mutableMapOf()
+    private val connections: MutableMap<ConnectionTag, Connection> = mutableMapOf()
 
     override fun getName(): String = "DatabaseBridge"
 
+    sealed class Connection {
+        class Connected(val driver: DatabaseDriver) : Connection()
+        class Waiting(var queueW: ArrayList<(() -> Unit)>) : Connection()
+
+        var queue: ArrayList<(() -> Unit)> = arrayListOf()
+            get() = when (this) {
+                is Connection.Connected -> arrayListOf()
+                is Connection.Waiting -> this.queueW
+            }
+    }
+
     @ReactMethod
-    fun setUp(
+    fun initialize(
         tag: ConnectionTag,
         databaseName: String,
-        schema: SQL,
         schemaVersion: Int,
         promise: Promise
     ) {
-        val driver = DatabaseDriver(
-                reactContext,
-                DatabaseDriver.Configuration(
-                        databaseName,
-                        schema,
-                        schemaVersion
-                )
-        )
-        connections[tag] = driver
-        promise.resolve(true)
+        assert(connections[tag] == null) { "A driver with tag $tag already set up" }
+        val promiseMap = Arguments.createMap()
+
+        try {
+            connections[tag] = Connection.Connected(
+                    driver = DatabaseDriver(
```

Comment:
I don't think that's needed. `Connection.Connected` takes only 1 arg and writing `driver = DatabaseDriver` is kinda repetitive. This is not Java 😝

Reply:
You are right about it not being Java - in Java I wouldn't be able to do something like this! 
I was sceptical about this at first but in the end I decided to implement it to a) increase readability of code, b) make it similar to the solution presented in Swift to decrease needed work when someone is working on both codebases later on

Reply:
👌